File: client/src/main/scala/stasis/client/ops/recovery/stages/internal/DestagedByteStringSource.scala
package stasis.client.ops.recovery.stages.internal

import java.nio.file.Path

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.control.NonFatal

import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.scaladsl.FileIO
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString
import org.apache.pekko.Done
import org.apache.pekko.NotUsed

import stasis.client.ops.Metrics
import stasis.client.ops.recovery.Providers

class DestagedByteStringSource(source: Source[ByteString, NotUsed]) {
  def destage(to: Path)(implicit providers: Providers, mat: Materializer): Future[Done] = {
    implicit val ec: ExecutionContext = mat.executionContext

    val metrics = providers.telemetry.metrics[Metrics.RecoveryOperation]

    providers.staging
      .temporary()
      .flatMap { staged =>
        source
          .wireTap(bytes => metrics.recordEntityChunkProcessed(step = "destaged", bytes = bytes.length))
          .runWith(FileIO.toPath(staged))
          .flatMap(_ => Future.successful(Done))
          .flatMap(_ => providers.staging.destage(from = staged, to = to))
          .recoverWith { case NonFatal(e) => providers.staging.discard(staged).flatMap(_ => Future.failed(e)) }
      }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/stages/internal/DestagedByteStringSource.kt
package stasis.client_android.lib.ops.recovery.stages.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okio.Source
import okio.buffer
import okio.sink
import stasis.client_android.lib.ops.recovery.Providers
import java.nio.file.Files
import java.nio.file.Path

object DestagedByteStringSource {
    suspend fun Source.destage(to: Path, providers: Providers) = withContext(Dispatchers.IO) {
        val staged = providers.staging.temporary()

        try {
            staged.sink().buffer().use {
                it.writeAll(this@destage)
            }

            providers.staging.destage(from = staged, to = to)
        } catch (e: Throwable) {
            providers.staging.discard(staged)
            throw e
        }
    }
}
